activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -94,7 +94,11 @@ def self.[](*args)
     #
     # This value can be later fetched using either +:key+ or +'key'+.
     def []=(key, value)
-      regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      if value.frozen?
+        regular_writer(convert_key(key), convert_value(value))
+      else
+        regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      end
```

Comment:
I initially followed that approach but @senny fixed the same in https://github.com/rails/rails/pull/10784, that would have a 'side effect' as described in the PR.